File: Java/src/deepspace/SpecificDamage.java
package deepspace;

import java.util.ArrayList; // import the ArrayList class

/**
 *
 * @author Mario Megías Mateo and Shao Jie Hu Chen.
 */
public class SpecificDamage extends Damage {
    
    private ArrayList<WeaponType> weapons;
    
    /**
     * @brief Constructor
     * @param s Number of shields
     * @param wl List of types of weapons
     */
    SpecificDamage(ArrayList<WeaponType> wl, int s) {
        super(s);
        if (wl != null) {
            weapons = new ArrayList<WeaponType>(wl);
        } else {
            weapons = new ArrayList<WeaponType>();
        }
    }
    
    /**
     * @brief Copies the implicit object. ABSTRACT.
     * @return A copy of the implicit object.
     */
    @Override
    public SpecificDamage copy() {
        return new SpecificDamage(weapons, getNShields());
    }
    
    /**
     * @brief UI Interface. Overrride.
     * @return Interface to UI
     */
    @Override
    SpecificDamageToUI getUIversion(){
        return new SpecificDamageToUI(this);
    }
    
    /**
     * @brief This method finds whether w contains t type.
     * @param w ArrayList of Weapons
     * @param t WeaponType
     * @return Position at array if found. Else, a negative number. 
     */
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t) {
        int pos = -1;
        boolean found = false;
        
        for (int i=0; i<w.size() && !found; ++i) {
            if (w.get(i).getType() == t) {
                pos = i;
                found = true;
            }
        }
        return pos;
    }
    
    /**
     * @brief It reduces the Damage according to restrictions provided by w 
     * and s. Override.
     * @param w ArrayList of Weapons to adjust.
     * @param s ArrayList of ShieldsBooster to adjust.
     * @return New instance of Damage, with adjusted restrictions.
     */
    @Override
    public SpecificDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int numberofshields = adjustShields(s);
        int pos = 0;
        ArrayList<Weapon> wCopy = new ArrayList<Weapon>(w);
        ArrayList<WeaponType> weaponsCopy = new ArrayList<WeaponType>(this.weapons);
        
        while (pos < weaponsCopy.size()) {
            WeaponType weapon = weaponsCopy.get(pos);
            int pos_encountered = this.arrayContainsType(wCopy, weapon);

            if (pos_encountered >= 0) {
                wCopy.remove(pos_encountered);
                ++pos;
            } else {
                weaponsCopy.remove(weapon); // first element found 
            }
        }
        
        return new SpecificDamage(weaponsCopy, numberofshields);
    }
    
    /**
     * @brief It discards a Weapon from the Damage. Override.
     * @param w
     */
    @Override
    public void discardWeapon(Weapon w) {
        weapons.remove(w.getType());
    }
    
    /**
     * @brief Establishes when a Damage is inoffensive.
     * @return True when there are not shields not weapons. False otherwise. 
     */
    @Override
    public boolean hasNoEffect() {
        return (super.hasNoEffect() && weapons.isEmpty());
    }
    
    /**
     * @brief Getter
     * @return weapons attribute
     */
    public ArrayList<WeaponType> getWeapons() {
        return new ArrayList<WeaponType>(weapons);
    }
    
    /**
     * @brief toString
     * @return A string with the information of the state of the object
     */
    @Override
    public String toString() {
        String cadSuper = super.toString();
        String info = "";
        info += cadSuper.substring(0, cadSuper.length()-1);
        info += ", weapons: " + weapons.toString();
        info += "]";
        return info;
    }
    
}

File: Ruby/SpecificDamage.rb
#encoding:utf-8

# +-------------------------------------------------------------------+ #
# | DEEPSPACE (Ruby)                                                  | #
# | Authors: Shao Jie Hu Chen and Mario Megías Mateo                  | #
# +-------------------------------------------------------------------+ #
# | Specific Damage.                                                  | #
# +-------------------------------------------------------------------+ #


module Deepspace

require_relative 'Damage.rb'
require_relative 'Weapon.rb'
require_relative 'ShieldBooster.rb'
require_relative 'SpecificDamageToUI.rb'

class SpecificDamage < Damage
    def initialize(wl,s)
        super(s)
        if wl != nil
            @weapons = Array.new(wl)
        else
            @weapons = Array.new
        end
    end

    public_class_method :new

    attr_reader :weapons

    # Override
    def copy
        return SpecificDamage.new(weapons, nShields)
    end

    # Override
    def getUIversion
        return SpecificDamageToUI.new(self)
    end

    def arrayContainsType(weapon_arr, type)
        index = -1
        count = 0
        found = false
        while count < weapon_arr.length && !found
            if (weapon_arr[count].type == type)
                index = count
                found = true
            else
                count += 1
            end
        end
        return index
    end

    private :arrayContainsType

    # Override
    def adjust(theWeaponsArray, theShieldsArray)
        nShieldsCopy = super

        weaponsCopy = Array.new(weapons)
        wCopy = Array.new(theWeaponsArray)
        index = 0

        while index < weaponsCopy.length
            weaponType = weaponsCopy[index]
            pos = arrayContainsType(wCopy,weaponType)
            if pos >= 0
                wCopy.delete_at(pos)
                index += 1
            else
                weaponsCopy.delete_at(index)
            end
        end

        return SpecificDamage.new(wCopy, nShieldsCopy)
    end

    # Override
    def discardWeapon(weapon)
        index = arrayContainsType(@weapons, weapon.type)

        if index >= 0
            @weapons.delete_at(index)
        end
    end

    # Override
    def hasNoEffect
        return (super && (weapons.length == 0))
    end

    # Override
    def to_s
        s = super
        s2 = s.sub("]", "")

        weaponsString = ""
        weapons.each do |w|
            weaponsString += "#{w.to_s}, "
        end
        
        s2 += "; weaponsType: #{weaponsString}]"

        return s2
    end

end # SpecificDamage
end # Deepspace

if $0 == __FILE__ then

    w1 = Deepspace::Weapon.new("Laser 1", Deepspace::WeaponType::LASER, 10)
    w2 = Deepspace::Weapon.new("Laser 2", Deepspace::WeaponType::LASER, 20)
    w3 = Deepspace::Weapon.new("Missile 1", Deepspace::WeaponType::MISSILE, 10)
    w4 = Deepspace::Weapon.new("Missile 2", Deepspace::WeaponType::MISSILE, 10)
    
    s1 = Deepspace::ShieldBooster.new("ShieldBooster 1", 10, 2)
    s2 = Deepspace::ShieldBooster.new("ShieldBooster 2", 20, 7)

    puts '-----------------------------------------------'
    puts 'Constructor'
    puts '-----------------------------------------------'
    specDamage = Deepspace::SpecificDamage.new([w1,w2,w3,w4], 10)
    puts specDamage.to_s

    puts "\n\n"

    puts '-----------------------------------------------'
    puts 'Copy'
    puts '-----------------------------------------------'
    copyDamage = specDamage.copy
    puts copyDamage.to_s

    puts "\n\n"

    puts '-----------------------------------------------'
    puts 'UI version'
    puts '-----------------------------------------------'
    uiDamage = specDamage.getUIversion
    puts uiDamage.to_s

    puts "\n\n"

    puts '-----------------------------------------------'
    puts 'Adjust'
    puts '-----------------------------------------------'
    
    adjustDamage = specDamage.adjust([w1], [s1,s2])

    puts adjustDamage.to_s
    puts specDamage.to_s

    puts "\n\n"

    
    
end # Test